In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

pd.options.plotting.backend = 'plotly'

import plotly.io as pio
pio.renderers.default = "notebook_connected"

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_excel('../../data/OptimizingTheGrowthENV/redLettuce/CASE9/dataset_20210504.xlsx')
df

시간  내부온도관측치    내부습도관측치  CO2관측치       EC관측치    외부온도관측치  \
0      2021-05-04 00:00     30.6  46.700001   589.0    1.091304  31.700001   
1      2021-05-04 00:01     30.6  46.700001   589.0    1.090534  31.799999   
2      2021-05-04 00:02     30.6  46.700001   592.0    1.092073  31.600000   
3      2021-05-04 00:03     30.6  46.700001   587.0    1.092843  31.700001   
4      2021-05-04 00:04     30.6  46.700001   590.0    1.091304  31.600000   
...                 ...      ...        ...     ...         ...        ...   
40300               NaN      NaN        NaN     NaN         NaN        NaN   
40301               NaN      NaN        NaN     NaN         NaN        NaN   
40302               NaN      NaN        NaN     NaN         NaN        NaN   
40303  2021-05-31 23:58     23.6   0.000000     0.0    0.000000  20.500000   
40304  2021-05-31 23:59     23.5   0.000000   201.0  132.000000  17.500000   

         외부습도관측치  펌프상태  펌프작동남은시간  최근분무량  ...  냉방온도  난방온도  기준온도    내외부온도차  \
0      23.600000   NaN       NaN    0.0  ...  21.0  20.0  20.5       NaN   
1      23.700001   0.0       0.0    0.0  ...  21.0  20.0  20.5       NaN   
2      23.700001   0.0       0.0    0.0  ...  21.0  20.0  20.5 -1.199999   
3      23.700001   0.0       0.0    0.0  ...  21.0  20.0  20.5       NaN   
4      23.700001   NaN       NaN    0.0  ...  21.0  20.0  20.5       NaN   
...          ...   ...       ...    ...  ...   ...   ...   ...       ...   
40300        NaN   NaN       NaN    NaN  ...   NaN   NaN   NaN       NaN   
40301        NaN   NaN       NaN    NaN  ...   NaN   NaN   NaN       NaN   
40302        NaN   NaN       NaN    NaN  ...   NaN   NaN   NaN       NaN   
40303        NaN   NaN       NaN    NaN  ...   NaN   NaN   NaN       NaN   
40304        NaN   NaN       NaN    NaN  ...   NaN   NaN   NaN       NaN   

       난방부하       냉방부하  총추정광량  백색광추정광량  적색광추정광량  청색광추정광량  
0       0.0  48.000002    0.0      0.0      0.0      0.0  
1       0.0  48.000002    0.0      0.0      0.0      0.0  
2       0.0  48.000002    0.0      0.0      0.0      0.0  
3       0.0  48.000002    0.0      0.0      0.0      0.0  
4       0.0  48.000002    0.0      0.0      0.0      0.0  
...     ...        ...    ...      ...      ...      ...  
40300   NaN        NaN    NaN      NaN      NaN      NaN  
40301   NaN        NaN    NaN      NaN      NaN      NaN  
40302   NaN        NaN    NaN      NaN      NaN      NaN  
40303   NaN        NaN    NaN      NaN      NaN      NaN  
40304   NaN        NaN    NaN      NaN      NaN      NaN  

[40305 rows x 39 columns]

In [3]:
df.columns

Index(['시간', '내부온도관측치', '내부습도관측치', 'CO2관측치', 'EC관측치', '외부온도관측치', '외부습도관측치',
       '펌프상태', '펌프작동남은시간', '최근분무량', '일간누적분무량', '냉방상태', '냉방작동남은시간', '난방상태',
       '난방작동남은시간', '내부유동팬상태', '내부유동팬작동남은시간', '외부환기팬상태', '외부환기팬작동남은시간',
       '화이트 LED상태', '화이트 LED작동남은시간', '화이트 LED동작강도', '레드 LED상태', '레드 LED작동남은시간',
       '레드 LED동작강도', '블루 LED상태', '블루 LED작동남은시간', '블루 LED동작강도', '카메라상태', '냉방온도',
       '난방온도', '기준온도', '내외부온도차', '난방부하', '냉방부하', '총추정광량', '백색광추정광량', '적색광추정광량',
       '청색광추정광량'],
      dtype='object')

In [4]:
# 결측치 확인
nan = df.isna().sum().to_frame(name='NaN')
nan

NaN
시간             11503
내부온도관측치        11514
내부습도관측치        11510
CO2관측치         11510
EC관측치          11513
외부온도관측치        11517
외부습도관측치        11517
펌프상태           17956
펌프작동남은시간       17956
최근분무량          11505
일간누적분무량        11505
냉방상태           13291
냉방작동남은시간       13292
난방상태           21424
난방작동남은시간       21425
내부유동팬상태        22929
내부유동팬작동남은시간    22930
외부환기팬상태        18008
외부환기팬작동남은시간    18008
화이트 LED상태      13003
화이트 LED작동남은시간  13004
화이트 LED동작강도    13004
레드 LED상태       13109
레드 LED작동남은시간   13109
레드 LED동작강도     13109
블루 LED상태       13233
블루 LED작동남은시간   13233
블루 LED동작강도     13233
카메라상태          22867
냉방온도           11995
난방온도           11995
기준온도           11995
내외부온도차         30763
난방부하           12000
냉방부하           12000
총추정광량          11996
백색광추정광량        11996
적색광추정광량        11996
청색광추정광량        11996

In [5]:
# 결측값 없는 columns
nan.loc[nan['NaN']==0].index

Index([], dtype='object')

In [6]:
# 결측치 있는 columns
nan.loc[nan['NaN']!=0].index

Index(['시간', '내부온도관측치', '내부습도관측치', 'CO2관측치', 'EC관측치', '외부온도관측치', '외부습도관측치',
       '펌프상태', '펌프작동남은시간', '최근분무량', '일간누적분무량', '냉방상태', '냉방작동남은시간', '난방상태',
       '난방작동남은시간', '내부유동팬상태', '내부유동팬작동남은시간', '외부환기팬상태', '외부환기팬작동남은시간',
       '화이트 LED상태', '화이트 LED작동남은시간', '화이트 LED동작강도', '레드 LED상태', '레드 LED작동남은시간',
       '레드 LED동작강도', '블루 LED상태', '블루 LED작동남은시간', '블루 LED동작강도', '카메라상태', '냉방온도',
       '난방온도', '기준온도', '내외부온도차', '난방부하', '냉방부하', '총추정광량', '백색광추정광량', '적색광추정광량',
       '청색광추정광량'],
      dtype='object')

# 
## <span style='color:#0000FF; font-family:cursive'> Time </span>

#### DataFrame에서 결측치 확인

In [7]:
# 시간이 결측인 경우, 모든 데이터가 결측임을 확인
time = df.loc[df['시간'].isna()]
time.isna().sum().to_frame(name='NaN')

NaN
시간             11503
내부온도관측치        11503
내부습도관측치        11503
CO2관측치         11503
EC관측치          11503
외부온도관측치        11503
외부습도관측치        11503
펌프상태           11503
펌프작동남은시간       11503
최근분무량          11503
일간누적분무량        11503
냉방상태           11503
냉방작동남은시간       11503
난방상태           11503
난방작동남은시간       11503
내부유동팬상태        11503
내부유동팬작동남은시간    11503
외부환기팬상태        11503
외부환기팬작동남은시간    11503
화이트 LED상태      11503
화이트 LED작동남은시간  11503
화이트 LED동작강도    11503
레드 LED상태       11503
레드 LED작동남은시간   11503
레드 LED동작강도     11503
블루 LED상태       11503
블루 LED작동남은시간   11503
블루 LED동작강도     11503
카메라상태          11503
냉방온도           11503
난방온도           11503
기준온도           11503
내외부온도차         11503
난방부하           11503
냉방부하           11503
총추정광량          11503
백색광추정광량        11503
적색광추정광량        11503
청색광추정광량        11503

###  
시간이 NaN인 경우의 행 모두 삭제  
  
⠀ →⠀ df의 행이 **40305** 에서 **28802** 로 줄었다.

In [8]:
df = df.dropna(subset=['시간'])
len(df)

28802